### ISSUE TITLE
Loss goes to NaN after several epochs

### ISSUE THREAD

**Issue by CharlieCheckpt** [bytedance/ibot#19]
Hello, 

First of all, well done and thank you for this great work.

I am trying to launch iBOT experiments but struggle with loss going to NaN after a few epochs. The training loss increases (see below), which seems weird. Have you faced similar issues during your experiments and if yes, how did you solve it ? 
I have limited ressources to launch my experiments, so I can't play too much on the parameters, and any tip would be greatly appreciated.

I know that using mixed precision, as this is the case here, can lead to stability issues. I indeed do not get any nan loss when setting `use_fp16` to False. But I'd rather let `use_fp16` to True to keep a reasonable training time.

I tried to increase eps of AdamW to 1e-6 and of Batch norm layers to 6.1e-5 (as proposed [here](https://github.com/pytorch/pytorch/issues/40497#issuecomment-1002372380)), but this did not work. 
I see in #17 that you propose to decrease beta opt 2 of AdamW, can you comment a bit more on this ? And on other techniques if you know any ?

Below is how I launched the training (I'm on a slurm cluster so I use `run_with_submitit.py` that calls `main_ibot.py` ).
Dataset size is ~4M images, hence around 3 times larger than ImageNet10K training set. This is why I divided by 3 the number of epochs for `warmup_teacher_temp_epochs` and `warmup_epochs`. Not sure this makes sense though.
```
python run_with_submitit.py --arch vit_small \
    --ngpus 4 \
    --nodes 4 \
    --num_workers 8 \
    --teacher_temp 0.07 \
    --warmup_teacher_temp_epochs 10 \
    --warmup_epochs 3 \
    --norm_last_layer false \
    --epochs 100 \
    --batch_size_per_gpu 112 \
    --shared_head true \
    --out_dim 8192 \
    --local_crops_number 10 \
    --global_crops_scale 0.25 1 \
    --local_crops_scale 0.05 0.25 \
    --pred_ratio 0 0.3 \
    --pred_ratio_var 0 0.2 \
    --timeout 1200 \
    --partition gpu_xxx \
    --data_path "xxx" \
    --saveckp_freq 1
```

Below, you can find the training metrics recorded :
```
{"train_loss": 6.99587931743066, "train_cls": 4.331486056964688, "train_patch": 2.6643932607626066, "train_lr": 0.0005831685964416835, "train_wd": 0.040029588545113536, "train_acc": 0.6155056208539762, "train_nmi": 0.1298466117645908, "train_ari": 0.006587329895082833, "train_fscore": 0.043441455577130375, "train_adjacc": -1, "epoch": 0}
{"train_loss": 10.466352438147872, "train_cls": 6.67679833335828, "train_patch": 3.789554104414066, "train_lr": 0.0017500000000000005, "train_wd": 0.040207159993842605, "train_acc": 0.408415272718953, "train_nmi": 0.14113477636477684, "train_ari": 0.0071149456743873594, "train_fscore": 0.030568697062104223, "train_adjacc": -1, "epoch": 1}
{"train_loss": 12.008997473409751, "train_cls": 8.066221890750954, "train_patch": 3.9427755813775858, "train_lr": 0.002916831403558317, "train_wd": 0.0405621652690049, "train_acc": 0.3335961760343232, "train_nmi": 0.13128954990810676, "train_ari": 0.004251988918558582, "train_fscore": 0.03003438171217746, "train_adjacc": -1, "epoch": 2}
```

Below you can find more information about the config used for the experiment :
```
act_in_head: gelu
arch: vit_small
batch_size_per_gpu: 112
clip_grad: 3.0
comment:
constraint: ""
data_path: ""
dist_url: env://
drop_path: 0.1
epochs: 100
freeze_last_layer: 1
global_crops_number: 2
global_crops_scale: [0.25, 1.0]
gpu: 0
lambda1: 1.0
lambda2: 1.0
local_crops_number: 10
local_crops_scale: [0.05, 0.25]
local_rank: 0
lr: 0.0005
min_lr: 1e-06
momentum_teacher: 0.996
ngpus: 4
nodes: 4
norm_in_head: None
norm_last_layer: False
num_workers: 8
optimizer: adamw
out_dim: 8192
output_dir: ""
partition: gpu_xxx
patch_out_dim: 8192
patch_size: 16
pred_ratio: [0.0, 0.3]
pred_ratio_var: [0.0, 0.2]
pred_shape: block
pred_start_epoch: 0
qos: qos_xxx
rank: 0
saveckp_freq: 1
seed: 0
shared_head: True
shared_head_teacher: True
teacher_patch_temp: 0.07
teacher_temp: 0.07
timeout: 1200
use_fp16: True
use_masked_im_modeling: True
warmup_epochs: 3
warmup_teacher_patch_temp: 0.04
warmup_teacher_temp: 0.04
warmup_teacher_temp_epochs: 10
weight_decay: 0.04
weight_decay_end: 0.4
window_size: 7
world_size: 16
```

Best regards

**Comment by CharlieCheckpt** [comment]
After looking at the code in `main_ibot.py` it seems that setting `norm_last_layer` to true may help, as well as decreasing `teacher_temp` to 0.04. Apologies for not reading the documentation carefully.
I launched a new job with these two changes, I will let you know the results.

**Comment by jensenzhoujh** [comment]
Hi @CharlieCheckpt,

Glad that you run your exps successfully. AdamW uses a default optim betas (0.9, 0.999), and you can try to decrease the second param until training is stable. In my experiments, I do not scale `warmup_teacher_temp_epochs` according to the data size which tends to result in better performance. You can take a look at the args in ImageNet-22K exps. This should not affect the stability though. In fact, I rarely encounter stability issues when conducting experiments with iBOT on ImageNet-1/22K. But I think it's sensible to tweak around args like ema momentum, center momentum, teacher_temp, etc with new datasets since these are major techniques to avoid model collapse in DINO.

**Comment by CharlieCheckpt** [comment]
Thank you for your answer @shallowtoil. As I mentioned above, I ran another experiment setting `norm_last_layer` to `true` and `teacher_temp` to `0.04`.  It seems to have solved the NaN issue I encountered. I have no problem for now at epoch 90, experiment is still running. One can find below the training curves for reference. They are not super stable but it seems the model is learning.

I will explore what you propose too.

<img width="1414" alt="image" src="https://user-images.githubusercontent.com/10373261/168644880-15a25464-e935-4140-9656-1a4bf2cc8ad6.png">

<img width="1397" alt="image" src="https://user-images.githubusercontent.com/10373261/168644745-7af1810a-7143-483d-aad4-93668545ced9.png">

<img width="1414" alt="image" src="https://user-images.githubusercontent.com/10373261/168645053-b05bc4bc-b156-4ef4-bf73-a7a5527625d9.png">

<img width="1414" alt="image" src="https://user-images.githubusercontent.com/10373261/168645101-545113fb-373a-4fec-92e8-f87ebf04ee2b.png">

<img width="1414" alt="image" src="https://user-images.githubusercontent.com/10373261/168645522-fb342a24-f43b-41c4-a460-c93f29ba45be.png">

<img width="1414" alt="image" src="https://user-images.githubusercontent.com/10373261/168645601-77d6f9d4-6c21-485a-a739-9189edcac30a.png">

Best

**Comment by jensenzhoujh** [comment]
Close this for now. Feel free to re-open it.
